log4j2/src/main/java/com/google/common/flogger/backend/log4j2/Log4j2SimpleMessageFormatter.java

```diff
@@ -0,0 +1,255 @@
+     * and can always be safely called if that method returned {@code false} for the same log data.
+     *
+     * @param logData the log statement data.
+     * @return the single logged value as a string.
+     * @throws IllegalStateException if the log data had arguments to be formatted (i.e. there was a
+     *     template context).
+     */
+    public static String getLiteralLogMessage(LogData logData) {
+        return MessageUtils.safeToString(logData.getLiteralArgument());
+    }
+
+    /**
+     * An internal helper method for logger backends which are aggressively optimized for performance.
+     * This method is a best-effort optimization and should not be necessary for most implementations.
+     * It is not a stable API and may be removed at some point in the future.
+     *
+     * <p>This method attempts to determine, for the given log data and log metadata, if the default
+     * message formatting performed by the other methods in this class would just result in the
+     * literal log message being used, with no additional formatting.
+     *
+     * <p>If this method returns {@code false} then the literal log message can be obtained via {@link
+     * #getLiteralLogMessage(LogData)}, otherwise {@code appendFormatted()} should be used instead.
+     *
+     * <p>By calling this class it is possible to more easily detect cases where using buffers to
+     * format the log message is not required. Obviously a logger backend my have its own reasons for
+     * needing buffering (e.g. prepending log site data) and those must also be taken into account.
+     *
+     * @param logData the log statement data.
+     * @param metadata the metadata intended to be formatted with the log statement.
+     * @param keysToIgnore a set of metadata keys which are known not to appear in the final formatted
+     *     message.
+     */
+    public static boolean mustBeFormatted(
+            LogData logData, MetadataProcessor metadata, Set<MetadataKey<?>> keysToIgnore) {
+        // If there are logged arguments or more metadata keys than can be ignored, we fail immediately
+        // which avoids the cost of creating the metadata key set (so don't remove the size check).
+        return logData.getTemplateContext() != null
+                || metadata.keyCount() > keysToIgnore.size()
+                || !keysToIgnore.containsAll(metadata.keySet());
+    }
+
+    /**
+     * Helper method to format the log message and any metadata with the default formatting. This is
+     * exposed in a public API by the default log message formatter (but also used in deprecated
+     * methods for now).
+     */
+    private static String format(LogData logData, MetadataProcessor metadata) {
+        return mustBeFormatted(logData, metadata, DEFAULT_KEYS_TO_IGNORE)
+                ? appendFormatted(logData, metadata, DEFAULT_HANDLER, new StringBuilder()).toString()
+                : getLiteralLogMessage(logData);
+    }
+
```

Comment:
Please don't copy and paste deprecated code like this. Only add code you are using.
Ideally this class wouldn't be needed at all (or should at least be almost trivial).

Reply:
removed it. sry for that.